### PR TITLE
Download Attachment changes.  Support for Edge and change to Blob

### LINKED ISSUE #295: Internet Explorer can't download files
## Steps to reproduce
1. Upload a file using the file upload method form PrivateBin
2. Try to download the file with Internet Explorer (10, 11, Edge)

### What happens
Nothing happens, the browser does not download the file.

### What should happen
A file download should be initiated

## Additional information
Tried with the latest version from PrivateBin I pulled from this repo.

## Basic information

**Server OS:**
CentOS 7

**Webserver:**
APACHE 2

**Browser:**
Internet explorer 10, 11, EDGE

**PrivateBin version:**
1.1.1

I can reproduce this issue on <https://privatebin.net>: No / File uploads disabled

### PR DESCRIPTION
**Pull request by otaku**
This PR fixes #295 

Internet Explorer uses msSaveBlob.  I've had to bind a click event however.  I also swapped to using the Blob for the download link on other browsers instead of using the dataURI.

### CODE REVIEW

**Comment by r4sas** [comment]
But why for all browsers?

**Comment by otaku** [comment]
It does check if Blob is a supported function before doing so, but the primary benefit is mousing over the download link doesn't attempt to display the really long dataURI.  This is very noticeable with a large attachment.

There however is a caveat with hindsight; createObjectURL will create a temporary file, which can be removed by revokeObjectURL, or when the browser decides to clean temporary files.  This could be fixed by adding a click event handler to call createObjectURL and a timeout to revoke after a certain amount of time.

**Comment by rugk** [comment]
> createObjectURL will create a temporary file

On disk? Really? I thought it only does this in the RAM…

In any case, first a big warning:
**Warning:** Blob URIs can cause security issues (XSS), as it happened with WhatsApp/Telegram [web clients](https://research.checkpoint.com/check-point-discloses-vulnerability-whatsapp-telegram/). So we would have to make sure one can never open this blob URI in a new tab or so.
I still need to re-read that attack there, but I feel better with data URIs for now in any case.

On the other hand blob URIs are not size-limited in contrast to data URIs.

In any case, we need to be careful here. Also, the base64 stuff you are doing at the beginning is quite hard to follow. Maybe needs some comments, at least? Do you really decode all that… mhh… does not sound too efficient, especially for large files.
But AFAIK `attachmentData` is always saved base64, so we would have to change the format there. 

----

All in all it also seems [IE supports data URIs](https://stackoverflow.com/questions/3916191/download-data-url-file/15832662#15832662). At least [Edge should](https://caniuse.com/#feat=datauri), even > 4GB.

So maybe we can search for better solutions.

**Comment by otaku** [comment]
@rugk not sure about the Blob stuff, let me know what you'd like to do.

The section you're referring to is where I decode a dataURI and transform it into a binary Blob.

DataURI format is data:[<mediatype>][;base64],<data>
I simply extract the media type and the data section after the comma.  The data section is passed through atob turning it from base64 to binary.

IE / Edge does not support dataURI to download, only though msSaveBlob or msSaveOrOpenBlob.  I can change it to where it only detects for Edge / IE and converts it to a Blob to be downloaded in that case.

**Comment by rugk** [comment]
>  The data section is passed through Base64.atob turning it from base64 to binary.

Yeah and [then](https://github.com/PrivateBin/PrivateBin/pull/298/files#diff-cf28a0bc390eee2482509aca8212f831R1937) you somehow get the number of the UTF-16 code for each character by iterating over the whole file.

> IE / Edge does not support dataURI to download, only though msSaveBlob or msSaveOrOpenBlob

Ah okay, only download.

Well… if that is really the only solution for Edge/IE, then go on with my review. I'll request that.

**Comment by otaku** [comment]
Relevant resources:
[MDN Data URI](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs)
[Microsoft Data URI](https://msdn.microsoft.com/en-us/library/cc848897%28v=vs.85%29.aspx)
[msSaveBlob](https://msdn.microsoft.com/en-us/library/hh772331(v=vs.85).aspx)

File: js/privatebin.js [hunk 1]
```diff
@@ -1971,6 +2001,7 @@ jQuery.PrivateBin = (function($, sjcl, Base64, RawDeflate) {
             me.hideAttachmentPreview();
             $attachmentLink.prop('href', '');
             $attachmentLink.prop('download', '');
+            $attachmentLink.unbind('click');
```
[the remark this answers is not in the document]
Comment (otaku) [reply js/privatebin.js]: There is no simple way to check if an element has a EventListener.  I could add a private variable to AttachmentViewer to track if ```$attachmentLink``` has the event handler.

**Pull request merged**